src/shogun/distributions/KernelDensity.cpp

```diff
@@ -91,12 +91,15 @@ bool KernelDensity::train(std::shared_ptr<Features> data)
 	return true;
 }
 
-SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<DenseFeatures<float64_t>>& test, int32_t leaf_size)
+SGVector<float64_t> KernelDensity::get_log_density(const std::shared_ptr<Features>& test, int32_t leaf_size)
 {
 	require(test,"data not supplied");
 
+	auto dense_feat = std::dynamic_pointer_cast<DenseFeatures<float64_t>>(test);
+	require(dense_feat, "Expected Features to be of type DenseFeatures.");
```

Comment:
@karlnapf is it fine to have this userfacing?

Reply:
can't this be moved to the crtp dispatcher? There is an error msg in there. But I guess it is some work. So then it is fine like this as well. Although people will not understand it. I also don't like using classnames in user-facing msgs (as they tend to change). What about say "dense real features"

Reply:
Hmm it would be some work, because this is not the train method. We would need a dispatcher for get_log_density?

Reply:
ah I see, maybe later right? :)

Reply:
indeed.. Otherwise there will be too much in this PR